Added sheets runbook (#1621)
mitodl/mitxpro: sheets/README.md

# xPro Google Sheets

**SECTIONS**
1. [What Is This?](#what-is-this)
1. [How It Works](#how-it-works)
1. [Types of Sheets](#types-of-sheets)
1. [Sheets Management](#sheets-management)
1. [Development Setup](#development-setup)

## What Is This?

We've had a few customer service-related use cases that we have needed to support. One option in terms of implementation
was to create several custom web UIs in xPro and give user access to all who needed it (CS representatives, finance 

> team members, etc.) This had the potential to become a persistent headache for developers, so we decided to support 

Reviewer: You can delete lines 12-14.

> these use cases through UIs that are very familiar to most users regardless of technical expertise: Google Sheets and

Reviewer: ```suggestion
Rather than support web UIs for internal interfaces, we chose to support Customer Support and Finance tasks through UIs that are very already familiar to them: Google Sheets and
```

be done, and it is automatically added to a form responses worksheet within a relevant spreadsheet.
    - That submission also appears in the main worksheet within the spreadsheet automatically via a query.
1. In response to the change in the spreadsheet, Google sends a file watch request to an endpoint in our app, which 
  indicates to us that the spreadsheet has been changed, so we may have some new requests to handle.
1. Our app scans the main worksheet of the spreadsheet for new or updated rows and takes the appropriate action if the
  data is all valid (e.g.: issuing a refund, creating new enrollment codes, etc.)
1. If a request was completed for some row, or if an error occurred, our app updates certain columns in the 
  main worksheet to indicate the status of that request.  

## Types of Sheets

Right now there are three sheet types:

1. Enrollment Code Request sheet (a.k.a. Coupon Request sheet)
1. Enrollment Code Assignment sheets (which are created based on Enrollment Code Request sheet entries)
1. Change of Enrollment Request sheet. This includes worksheets for the following:
    1. Refund Requests

>     1. Deferral Requests

Reviewer: Do you think it's OK to link to the sheets here? They do require authentication.

#### Enrollment Code Request Sheet & Assignment Sheets

_Details to be filled in later..._

#### Enrollment Change Requests (Refunds, Deferrals)

_Details to be filled in later..._

## Sheets Management

One cardinal rule to follow for all of the "request"-type sheets: **Do not edit the main worksheet directly, unless the 
columns exist specifically for user input**. Examples of columns that are intended for user input are the Finance 
columns in the Change of Enrollment sheets.

The terms "main worksheet" and "form responses worksheet" are used below. Here are some
screenshots for explanation.

**Enrollment Code Request worksheet tabs:**
![Enrollment Code Request worksheet tabs](images/enrollment-code-worksheet-tabs.png)

**Change of Enrollment Request worksheet tabs:**
![Change of Enrollment Request worksheet tabs](images/change-of-enrollment-worksheet-tabs.png)

> 

Reviewer: I think this section could use a little into, to make it clear that when there are errors, there are two options for resolution: editing and ignoring. 

After that I think it will be easier to follow.

> # xPro Google Sheets

Reviewer 2: ```suggestion
# xPRO Google Sheets
```

**SECTIONS**
1. [What Is This?](#what-is-this)
1. [How It Works](#how-it-works)
1. [Types of Sheets](#types-of-sheets)
1. [Sheet Basics](#sheet-basics)
1. [Handling Request Sheet Errors](#handling-request-sheet-errors)
1. [Handling Assignment Sheet Errors](#handling-assignment-sheet-errors)
1. [Development Setup](#development-setup)

## What Is This?

Rather than support web UIs for certain internal Customer Support and Finance tasks, we chose to support these tasks  
through UIs that are already very familiar to these teams: Google Sheets and Google Forms. 
Using various Google Drive and Sheets APIs, the following is now possible: 

1. CS/finance users can make various requests (e.g.: enrollment code creation/assignment, enrollment deferrals) by filling 
  out a Google Form or entering data into Google Sheets.
1. Those requests are automatically processed by our app.
1. Users can monitor the status of their requests directly in the spreadsheets.  

## How It Works

> The basic workflow for most flavors of xPro sheets is roughly as follows:

Reviewer 2: ```suggestion
The basic workflow for most flavors of xPRO sheets is roughly as follows:
```

    - That submission also appears in the main worksheet within the spreadsheet automatically via a query.
1. In response to the change in the spreadsheet, Google sends a file watch request to an endpoint in our app, which 
  indicates to us that the spreadsheet has been changed, so we may have some new/updated requests to handle.
1. Our app scans the main worksheet of the spreadsheet for new/updated rows and takes the appropriate action if the
  data is all valid (e.g.: issuing a refund, creating new enrollment codes, etc.)
1. If a request was completed for some row, or if an error occurred, our app updates certain columns in the 
  main worksheet to indicate the status of that request.

## Types of Sheets

Right now there are three sheet types:

1. Enrollment Code Request sheet (a.k.a. Coupon Request sheet)
1. Enrollment Code Assignment sheets (which are created based on Enrollment Code Request sheet entries)
1. Change of Enrollment Request sheet. This includes worksheets for the following:
    1. Refund Requests
    1. Deferral Requests

New requests for the "request"-type sheets are submitted via Google Form:

> 1. [Enrollment Code Request form](https://docs.google.com/forms/d/1XfV5ppOYCFHd2Z1B1uL75C-ChgytHfkTD8YOSQM0vFM/edit) 
> 1. [Refund Request form](https://docs.google.com/forms/d/1n-TYFd91zKkSkjjudoACMId-mfUeAQ-rGMCzTKqPVFQ/edit)
> 1. [Deferral Request form](https://docs.google.com/forms/d/1lEbKGfLxMwh0Drj-CQJo5QXU2evEqxEP5iFGB4qNrw8/edit)

Reviewer: @briangrossman alerted me that these forms _cannot_ be set to require authentication. So I think we should remove the links from here. 

```suggestion
1. Enrollment Code Request form
1. Refund Request form
1. Deferral Request form
```

**How to do it:** Find the row in the relevant "Form Responses" worksheet that matches the row in the main 
spreadsheet and directly edit the columns there. 

Example scenario: On the Enrollment Code Request spreadsheet, some row has some detail in the Errors column indicating that a
user doesn't exist due to a misspelled email address. To fix this, you would go to the "Form Responses" worksheet 
(which holds the the raw response data from the form), find the row that matches the one you saw on the main worksheet
with the error text, and updated the email column value there. 

Google sends file watch requests to our app if anything in the spreadsheet is edited, so those changes to the Form
Responses data should be handled automatically.

#### Ignoring Rows 

**Request rows cannot be deleted, only set to "ignored".** There are a couple different reasons for this. 
 
**When you'll want to set a row to ignored:** The request is no longer relevant (for example, the request is an accidental repeat). 

**How to do it:** Find the row in the relevant Form Responses worksheet that matches the

> row on the main worksheet** (check the "Editing Rows" section for details), **and enter `=TRUE` into the "Ignored?" 

Reviewer 2: Not sure what you're trying to highlight here with the **'s

Author: @briangrossman That's [Markdown](https://www.markdownguide.org/getting-started/) notation. Double-*'s style the text as bold. Have a look at the rendered version: https://github.com/mitodl/mitxpro/blob/1563049a7cf4da92a2b3c7e8933b36a5a7e56059/sheets/README.md#ignoring-rows
